source/_posts/2017/2017-02-15-minimalistic-way-to-create-your-first-nette-extension.md

```diff
@@ -0,0 +1,209 @@
+
+2. **By Reading extension of open-source packages**. They are similar to documentation: many features on various use cases I didn't understand yet.
+
+That lead to **overstretching my brain muscle**. It's like trying to jump over huge hole before even walking.
+
+
+### It made me think: "Could it be simpler?"
+
+What is essential purpose of the extension? It registers services to Nette Service Container.
+
+- Register services to Nette Container?
+- Register services?
+- **Register 1 service** - that's the one and only step we'll make today.
+
+So next time you'll think "gosh, this is so hard, I don't understand it, I'm so slow/lazy/...", stop for a moment and carefully look at the problem. There might be an easier way.
+
+**And now to the code!**
+
+
+## Register service in Nette Sandbox
+
+I consider [Nette Sandbox](https://github.com/nette/sandbox) the best way to show learn any Nette feature. Let's use it.
+
+If you want to register a service, what will you do?
+
+```yaml
+# app/config/config.neon
+
+services:
+    - App\Model\UserManager # put it there
+```
+
+File `app/config/config.neon` is like a socket.
+
+<img src="/assets/images/posts/2017/nette-extension/single-socket.jpg" class="thumbnail">
+
+There is **one place to active your computer**, when you plug it in.
+
+But what if your want to **plug in computer and mobile charger** in the same time?
+
+<img src="/assets/images/posts/2017/nette-extension/multi-socket.jpg" class="thumbnail">
+
+To load more services, we use the same interface as `app/config/config.neon`: **a file with service section that lists services**.
+
+
+## Create Local Extension in 5 Steps
+
+Let's say we want to decouple a FileSystem utilities.
+
+### 1. Pick a Name for Directory
+
+What about "FileSystem"? If you agree, create `src/FileSystem` directory.
+We will put configuration (sevices.neon) and classes there.
+
+
+### 2. Create or move Related Classes there
+
+Starting small, one service will do. When it grows, we can decouple it later.
+
+```php
+# src/FileSystem/FileSystem.php
+
+<?php declare(strict_types=1);
+
+namespace FileSystem;
+
+final class FileSystem
+{
+    // some awesome methods!
+}
+
+```
+
+### 3. Create config
+
+This is similar to `app/config/services.neon`, just in different location:
+
+```yaml
+# src/FileSystem/config/services.neon
+
+services:
+    - FileSystem\FileSystem
+```
+
+### 4. Create an Extension
+
+```php
+# src/FileSystem/DI/FileSystemExtension.php
+
+<?php declare(strict_types=1);
+
+namespace FileSystem\DI;
+
+use Nette\DI\Compiler;
+use Nette\DI\CompilerExtension;
+
+final class FileSystemExtension extends CompilerExtension
+{
+    public function loadConfiguration()
+    {
+        // this method loads servcies from config and registers them do Nette\DI Container
+        Compiler::loadDefinitions(
+            $this->getContainerBuilder(),
+            $this->loadFromFile(__DIR__.'/../config/services.neon')['services']
```

Comment:
Jo, tohle je trochu WTF. Přišlo to s Nette 2.4. 

Isset je takový nejistý a nabízí 2 možnosti na jednu věc.

Jak to zjendodušit?

Tohle je zase trochu WTF, že to je jiné než config. Ale kdyby se to jmenovalo `services.neon`, tak by to mohlo implikovat, že jde o sekci "services" (asi tak je to zamýšleno).

```yaml
# services.neon

- SomeService
```

Co myslíš,